nengo/neurons.py

```diff
@@ -727,7 +814,106 @@ def step_math(self, dt, J, spiked, voltage, recovery):
         recovery[spiked > 0] = recovery[spiked > 0] + self.reset_recovery
 
 
-class NeuronTypeParam(Parameter):
-    def coerce(self, instance, neurons):
-        self.check_type(instance, neurons, NeuronType)
-        return super().coerce(instance, neurons)
+class _Spiking(NeuronType):
+    """Abstract base class for neuron models that turn rate models into spiking ones"""
+
+    base_type = NeuronTypeParam("base_type")
+    amplitude = NumberParam("amplitude", low=0, low_open=True)
+
+    def __init__(self, base_type, amplitude=1.0):
+        super().__init__()
+
+        self.base_type = base_type
+        if "rates" not in base_type.probeable:
+            raise ValidationError(
+                "Must be a rate neuron type", attr="base_type", obj=self
+            )
+
+        self.negative = base_type.negative
+        self.amplitude = amplitude
+
+    @property
+    def probeable(self):
+        base = list(self.base_type.probeable)
+        base[base.index("rates")] = "spikes"
+        return tuple(base)
+
+    def gain_bias(self, max_rates, intercepts):
+        return self.base_type.gain_bias(max_rates, intercepts)
+
+    def max_rates_intercepts(self, gain, bias):
+        return self.base_type.max_rates_intercepts(gain, bias)
+
+    def rates(self, x, gain, bias):
+        return self.base_type.rates(x, gain, bias)
+
+
+class RegularSpiking(_Spiking):
+    """Turn a rate neuron model into a spiking one with regular inter-spike intervals.
+
+    Spikes at regular intervals based on the rates of the base neuron type.
```

Comment:
Could also add a reference to https://arxiv.org/abs/2002.03553 assuming `np.trunc` is changed to `np.floor`?

Reply:
Sure, go for it.